### PR TITLE
Add support for Deserialization's `:only` parameter for undercased symbols to be consistent with strong_parameters

### PR DESCRIPTION
**Pull request by NullVoxPopuli**
#### Purpose

Currently, you need to provide hyphenated strings for the only list, per the JSON API standard:
```
  result = ActiveModelSerializers::Deserialization
      .jsonapi_parse(params,
        only: [
          'name', 'short-description', 'domain',
          'starts-at', 'ends-at',
          'mail-payments-end-at', 'electronic-payments-end-at',
          'location', 'show-on-public-calendar',
          'registration-email-disclaimer'
      ])
```

This is inconsistent with rails' parameter keys, which are `:underscored_symbols`

#### Changes

Adds a test for various key formats to the `only` parameter, and supports `:underscored_symbols`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,13 @@
 ## 0.10.x
 
+Features:
+- [#1674](https://github.com/rails-api/active_model_serializers/pull/1674) Make Deserialization support underscored symbols (@NullVoxPopuli)
+
+Fixes:
+
+Misc:
+- [#1673](https://github.com/rails-api/active_model_serializers/pull/1673) Adds "How to" guide on using AMS with POROs (@DrSayre)
```
Comment (NullVoxPopuli) [inline CHANGELOG.md]: missed that this was under the wrong version when I merged this PR

File: Gemfile [hunk 2]
```diff
@@ -42,6 +42,10 @@ group :bench do
 end
 
 group :test do
+  # allows for rspec-like invocation of single tests
+  # e.g.: $ m test/path.rb:lineNum
+  gem 'm', '~> 1.4.2'
```
Comment (NullVoxPopuli) [inline Gemfile]: this is super handy!!! when I was testing the deserialization stuff, I just ran:

```
m test/action_controller/json_api/deserialization_test.rb:88
```

File: test/action_controller/json_api/deserialization_test.rb [hunk 3]
```diff
@@ -53,7 +62,33 @@ def test_deserialization_of_relationship_only_object
           assert_equal(expected, response)
         end
 
-        def test_deserialization
+        test 'deserialization with the only filter' do
```
Comment (NullVoxPopuli) [inline test/action_controller/json_api/deserialization_test.rb]: decided to convert to the nicer test name format while I was in here

**Comment by beauby** [comment]
Could you link to where exactly in the spec it says that? I just searched for "hyphen" and did not find anything related to your claim.
Moreover, ruby symbols can contain hyphens (`:'short-description'`).

**Comment by NullVoxPopuli** [comment]
I guess more accurately, the only list casing must match the input JSON API document's casing. Since I'm using hyphens, I had to use hyphenated only lists. Not a fan :-)

File: Gemfile (hunk 2 above)
Comment (bf4) [inline Gemfile]: lulz, I have that in my `Gemfile.local`, which is why I added that part to the Gemfile. also where I configure pry and byebug

Comment (bf4) [inline Gemfile]: ✂️ ?

Comment (NullVoxPopuli) [inline Gemfile]: idk, if we all use it while testing, shouldn't we keep it?

Comment (bf4) [inline Gemfile]: well, if we add it, we want to exclude it on CI.. one less thing to build.  and I haven't been using it lately. .ruby -Ilib:test path/to/file_test.rb

Comment (bf4) [inline Gemfile]: let's move to another PR

Comment (NullVoxPopuli) [inline Gemfile]: is it that much a hit to test time to include it? it's a super small gem.

Comment (bf4) [inline Gemfile]: It's a critical path, so yes

Comment (NullVoxPopuli) [inline Gemfile]: aight, I'll remove it

**Comment by NullVoxPopuli** [comment]
@bf4 here are the bin/bench results -- not sure how to interpret these yet

```
$ bin/bench
camel 315.38880611037115/ips; 3569 objects
camel_lower 408.1705834490325/ips; 2487 objects
dash 2913.0752016183824/ips; 719 objects
unaltered 6916944.973512944/ips; 1 objects
underscore 4844.546682389138/ips; 313 objects
Benchmark results:
{
  "commit_hash": "4c534e2",
  "version": "0.10.0.rc5",
  "rails_version": "4.2.4",
  "benchmark_run[environment]": "2.3.0p0",
  "runs": [
    {
      "benchmark_type[category]": "camel",
      "benchmark_run[result][iterations_per_second]": 315.389,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 3569
    },
    {
      "benchmark_type[category]": "camel_lower",
      "benchmark_run[result][iterations_per_second]": 408.171,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 2487
    },
    {
      "benchmark_type[category]": "dash",
      "benchmark_run[result][iterations_per_second]": 2913.075,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 719
    },
    {
      "benchmark_type[category]": "unaltered",
      "benchmark_run[result][iterations_per_second]": 6916944.974,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 1
    },
    {
      "benchmark_type[category]": "underscore",
      "benchmark_run[result][iterations_per_second]": 4844.547,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 313
    }
  ]
}
caching on: caching serializers: gc off 544.2550165880999/ips; 1831 objects
caching off: caching serializers: gc off 541.7677630965736/ips; 1831 objects
caching on: non-caching serializers: gc off 900.0483611458204/ips; 1242 objects
caching off: non-caching serializers: gc off 942.7729877809759/ips; 1242 objects
Benchmark results:
{
  "commit_hash": "4c534e2",
  "version": "0.10.0.rc5",
  "rails_version": "4.2.4",
  "benchmark_run[environment]": "2.3.0p0",
  "runs": [
    {
      "benchmark_type[category]": "caching on: caching serializers: gc off",
      "benchmark_run[result][iterations_per_second]": 544.255,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 1831
    },
    {
      "benchmark_type[category]": "caching off: caching serializers: gc off",
      "benchmark_run[result][iterations_per_second]": 541.768,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 1831
    },
    {
      "benchmark_type[category]": "caching on: non-caching serializers: gc off",
      "benchmark_run[result][iterations_per_second]": 900.048,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 1242
    },
    {
      "benchmark_type[category]": "caching off: non-caching serializers: gc off",
      "benchmark_run[result][iterations_per_second]": 942.773,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 1242
    }
  ]
}

```

**Comment by NullVoxPopuli** [comment]
Here are the bin/bench results from master for comparison on my machine

```
$ bin/bench
camel 329.74491332354927/ips; 3569 objects
camel_lower 431.3347129582817/ips; 2487 objects
dash 3128.0858195564238/ips; 719 objects
unaltered 7340100.331977236/ips; 1 objects
underscore 5448.019856354929/ips; 313 objects
Benchmark results:
{
  "commit_hash": "afe3c93",
  "version": "0.10.0.rc5",
  "rails_version": "4.2.4",
  "benchmark_run[environment]": "2.3.0p0",
  "runs": [
    {
      "benchmark_type[category]": "camel",
      "benchmark_run[result][iterations_per_second]": 329.745,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 3569
    },
    {
      "benchmark_type[category]": "camel_lower",
      "benchmark_run[result][iterations_per_second]": 431.335,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 2487
    },
    {
      "benchmark_type[category]": "dash",
      "benchmark_run[result][iterations_per_second]": 3128.086,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 719
    },
    {
      "benchmark_type[category]": "unaltered",
      "benchmark_run[result][iterations_per_second]": 7340100.332,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 1
    },
    {
      "benchmark_type[category]": "underscore",
      "benchmark_run[result][iterations_per_second]": 5448.02,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 313
    }
  ]
}
caching on: caching serializers: gc off 537.750254440679/ips; 1831 objects
caching off: caching serializers: gc off 553.4460087551487/ips; 1831 objects
caching on: non-caching serializers: gc off 894.1171422455924/ips; 1242 objects
caching off: non-caching serializers: gc off 867.697517396726/ips; 1242 objects
Benchmark results:
{
  "commit_hash": "afe3c93",
  "version": "0.10.0.rc5",
  "rails_version": "4.2.4",
  "benchmark_run[environment]": "2.3.0p0",
  "runs": [
    {
      "benchmark_type[category]": "caching on: caching serializers: gc off",
      "benchmark_run[result][iterations_per_second]": 537.75,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 1831
    },
    {
      "benchmark_type[category]": "caching off: caching serializers: gc off",
      "benchmark_run[result][iterations_per_second]": 553.446,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 1831
    },
    {
      "benchmark_type[category]": "caching on: non-caching serializers: gc off",
      "benchmark_run[result][iterations_per_second]": 894.117,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 1242
    },
    {
      "benchmark_type[category]": "caching off: non-caching serializers: gc off",
      "benchmark_run[result][iterations_per_second]": 867.698,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 1242
    }
  ]
}

```

**Comment by NullVoxPopuli** [comment]
though, my machine is doing other stuff... so I don't know how accurate the comparison is

**Comment by NullVoxPopuli** [comment]
Ok, now comparing the use of lazy.

With:
```
$ bin/bench -p transform
camel 311.8762299330334/ips; 3569 objects
camel_lower 440.19463572552354/ips; 2487 objects
dash 3126.2114551302902/ips; 719 objects
unaltered 6865572.452548717/ips; 1 objects
underscore 5634.915813540613/ips; 313 objects
Benchmark results:
{
  "commit_hash": "0f74747",
  "version": "0.10.0.rc5",
  "rails_version": "4.2.4",
  "benchmark_run[environment]": "2.3.0p0",
  "runs": [
    {
      "benchmark_type[category]": "camel",
      "benchmark_run[result][iterations_per_second]": 311.876,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 3569
    },
    {
      "benchmark_type[category]": "camel_lower",
      "benchmark_run[result][iterations_per_second]": 440.195,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 2487
    },
    {
      "benchmark_type[category]": "dash",
      "benchmark_run[result][iterations_per_second]": 3126.211,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 719
    },
    {
      "benchmark_type[category]": "unaltered",
      "benchmark_run[result][iterations_per_second]": 6865572.453,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 1
    },
    {
      "benchmark_type[category]": "underscore",
      "benchmark_run[result][iterations_per_second]": 5634.916,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 313
    }
  ]
}

```

without: 

```
$ bin/bench -p transform
camel 333.60132280231744/ips; 3569 objects
camel_lower 437.4532598251732/ips; 2487 objects
dash 3037.099781712828/ips; 719 objects
unaltered 7164712.048329625/ips; 1 objects
underscore 5449.851167410945/ips; 313 objects
Benchmark results:
{
  "commit_hash": "0f74747",
  "version": "0.10.0.rc5",
  "rails_version": "4.2.4",
  "benchmark_run[environment]": "2.3.0p0",
  "runs": [
    {
      "benchmark_type[category]": "camel",
      "benchmark_run[result][iterations_per_second]": 333.601,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 3569
    },
    {
      "benchmark_type[category]": "camel_lower",
      "benchmark_run[result][iterations_per_second]": 437.453,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 2487
    },
    {
      "benchmark_type[category]": "dash",
      "benchmark_run[result][iterations_per_second]": 3037.1,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 719
    },
    {
      "benchmark_type[category]": "unaltered",
      "benchmark_run[result][iterations_per_second]": 7164712.048,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 1
    },
    {
      "benchmark_type[category]": "underscore",
      "benchmark_run[result][iterations_per_second]": 5449.851,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 313
    }
  ]
}
```

File: lib/active_model_serializers/adapter/json_api/deserialization.rb [hunk 4]
```diff
@@ -140,12 +140,18 @@ def validate_payload(payload)
         # rubocop:enable Metrics/CyclomaticComplexity
 
         # @api private
-        def filter_fields(fields, options)
+        def filter_fields(hash, options)
+          hash = transform_keys(hash, options)
```
Comment (bf4) [inline lib/active_model_serializers/adapter/json_api/deserialization.rb]: if we're to transform the keys for attributes and relationships, then it should happen before this method is called, rather than within it

Comment (NullVoxPopuli) [inline lib/active_model_serializers/adapter/json_api/deserialization.rb]: So, `parse` should be like this?

```ruby
        # Same as parse!, but returns an empty hash instead of raising InvalidDocument
        # on invalid payloads.
        def parse(document, options = {})
          document = document.dup.permit!.to_h if document.is_a?(ActionController::Parameters)

          validate_payload(document) do |invalid_document, reason|
            yield invalid_document, reason if block_given?
            return {}
          end

          primary_data = document['data']
          attributes = primary_data['attributes'] || {}
          attributes['id'] = primary_data['id'] if primary_data['id']
          relationships = primary_data['relationships'] || {}

          attributes = transform_keys(attributes, options)
          relationships = transform_keys(relationships, options)

          attributes = filter_fields(attributes, options)
          relationships = filter_fields(relationships, options)

          hash = {}
          hash.merge!(parse_attributes(attributes, options))
          hash.merge!(parse_relationships(relationships, options))

          hash
        end
```

after removing transform_keys from `filter_fields`

it does make sense from a single-responsibility point of view -- just clarifying

Comment (bf4) [inline lib/active_model_serializers/adapter/json_api/deserialization.rb]: so, this diff should be something like

```diff
        # @api private
        def filter_fields(fields, options)
          if (only = options[:only])
-            only = Array(only).map { |o| o.to_s }
+            only = Array(only).map { |o| o.to_s.underscore }
-            fields.slice!(*only)
+            fields.select { |k, _| only.include?(k) }
          elsif (except = options[:except])
-            except = Array(except).map { |o| o.to_s }
+            except = Array(except).map { |o| o.to_s.underscore }
-            fields.except!(*except)
+            fields.reject { |k, _v| except.include?(k) }
+          else
+            fields
+          end
        end
```

Comment (NullVoxPopuli) [inline lib/active_model_serializers/adapter/json_api/deserialization.rb]: aight, I'll re-push with new bin/bench.

is adding the `.lazy` worth it?

Comment (NullVoxPopuli) [inline lib/active_model_serializers/adapter/json_api/deserialization.rb]:
```
$ bin/bench -p transform
^[[A^[[Bcamel 327.7302829996124/ips; 3569 objects
camel_lower 413.67373534184645/ips; 2487 objects
dash 3022.052284864508/ips; 719 objects
unaltered 6631231.042600424/ips; 1 objects
underscore 5613.2806073106685/ips; 313 objects
Benchmark results:
{
  "commit_hash": "0f74747",
  "version": "0.10.0.rc5",
  "rails_version": "4.2.4",
  "benchmark_run[environment]": "2.3.0p0",
  "runs": [
    {
      "benchmark_type[category]": "camel",
      "benchmark_run[result][iterations_per_second]": 327.73,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 3569
    },
    {
      "benchmark_type[category]": "camel_lower",
      "benchmark_run[result][iterations_per_second]": 413.674,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 2487
    },
    {
      "benchmark_type[category]": "dash",
      "benchmark_run[result][iterations_per_second]": 3022.052,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 719
    },
    {
      "benchmark_type[category]": "unaltered",
      "benchmark_run[result][iterations_per_second]": 6631231.043,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 1
    },
    {
      "benchmark_type[category]": "underscore",
      "benchmark_run[result][iterations_per_second]": 5613.281,
      "benchmark_run[result][total_allocated_objects_per_iteration]": 313
    }
  ]
}
```

File: lib/active_model_serializers/adapter/json_api/deserialization.rb [hunk 5]
```diff
@@ -93,8 +93,11 @@ def parse(document, options = {})
           attributes['id'] = primary_data['id'] if primary_data['id']
           relationships = primary_data['relationships'] || {}
 
-          filter_fields(attributes, options)
-          filter_fields(relationships, options)
+          attributes = transform_keys(attributes, options)
+          relationships = transform_keys(relationships, options)
```
Comment (bf4) [inline lib/active_model_serializers/adapter/json_api/deserialization.rb]: so, this is about transforming keys to the Rails internal format of `underscore`.

Comment (bf4) [inline lib/active_model_serializers/adapter/json_api/deserialization.rb]: This should then happen to `document`, right?

Comment (NullVoxPopuli) [inline lib/active_model_serializers/adapter/json_api/deserialization.rb]: yeah, which it is by default.
transform_keys has ` options[:key_transform] || :underscore`

So, I guess if options[key_transform] is set, maybe I should be calling transform_keys on the only/except data as well

Comment (bf4) [inline lib/active_model_serializers/adapter/json_api/deserialization.rb]: and once that is done, why should there be any change in how `only` and `unless` work.  I mean, transform has already `underscore`d everything..

Comment (NullVoxPopuli) [inline lib/active_model_serializers/adapter/json_api/deserialization.rb]: though, that would look wierd

            parsed_hash = ActiveModelSerializers::Deserialization.jsonapi_parse(
              params,
              only: [:image_height, :image_width],
              key_transform: :camel_lower
            )
            render json: parsed_hash

Comment (bf4) [inline lib/active_model_serializers/adapter/json_api/deserialization.rb]: > , maybe I should be calling transform_keys on the only/except data as well

No, that's used internally by us (Rails), so once we've done the key transform on the request data, then the `only` options should be using the *internal* casing, which in Rails is underscore

Comment (NullVoxPopuli) [inline lib/active_model_serializers/adapter/json_api/deserialization.rb]: cool

Comment (bf4) [inline lib/active_model_serializers/adapter/json_api/deserialization.rb]: Given `payload = { data: { id: 1, type: user, attributes: { crazy-name: 'Lou' } } }` 
Then ` ActiveModelSerializers::Deserialization.jsonapi_parse(payload, only: [:crazy_name])` will make the following transforms

1. `{ data: { id: 1, type: user, attributes: { crazy_name: 'Lou' } } }` 
2. `{ id: 1, crazy_name: 'Lou' }`
3. `{crazy_name: 'Lou' }`

**Comment by beauby** [comment]
Re: [comparing use of lazy](https://github.com/rails-api/active_model_serializers/pull/1674#issuecomment-208104877): using lazy there makes no sense whatsoever as 1. no transformation is being performed on the collection and 2. we want to iterate over the whole thing anyways.

File: lib/active_model_serializers/adapter/json_api/deserialization.rb (hunk 5 above)
Comment (beauby) [inline lib/active_model_serializers/adapter/json_api/deserialization.rb]: Note that you can currently explicitly give "translations" for every field, which would solve your issue. I'm not sure making it easy to diverge from the spec's recommendations is a good idea (I know ember does it, but that's not a reason).

**Comment by NullVoxPopuli** [comment]
That was my hunch

**Comment by NullVoxPopuli** [comment]
any other thoughts on this?

**Comment by beauby** [comment]
@NullVoxPopuli: https://github.com/rails-api/active_model_serializers/pull/1674#discussion_r59531514

**Comment by NullVoxPopuli** [comment]
@beauby  I removed it before your comment :-)

**Comment by beauby** [comment]
No, I meant:
> Note that you can currently explicitly give "translations" for every field, which would solve your issue. I'm not sure making it easy to diverge from the spec's recommendations is a good idea (I know ember does it, but that's not a reason).

**Comment by NullVoxPopuli** [comment]
oh, how can you explicitly give translations for each field?

**Comment by beauby** [comment]
Through the `keys` option: https://github.com/NullVoxPopuli/active_model_serializers/blob/underscored_keys_support_for_deserialization/lib/active_model_serializers/adapter/json_api/deserialization.rb#L66

**Comment by beauby** [comment]
So in your case `keys: { :'short-description' => :short_description }`

**Comment by NullVoxPopuli** [comment]
if that could be used as `only` instead of in addition to, I think that would be worth it.

Some of my objects have > 15 fields, so specifying anything more than what I need isn't fun

**Comment by beauby** [comment]
You could do something like
```ruby
translations = {
  :'foo-bar' => :foo_bar,
  :'foo-baz' => :foo_baz
}

only = translations.keys
```

**Comment by NullVoxPopuli** [comment]
what about unknown keys, like something you'd have in metadata?

**Comment by beauby** [comment]
Not sure I understand what you mean by unknown keys.

**Comment by beauby** [comment]
Back to this one, I realize I misunderstood the initial purpose of this PR.
What you're saying is that when deserializing, we should refer to the transformed keys when filtering, rather than the pristine keys present in the payload. That's clearly a possibility, although the latter allows for not earlier filtering, thus not transforming keys when not needed (and string manipulations usually account for quite a bit of processing time usually, that's why I went the second way).

**Comment by NullVoxPopuli** [comment]
I'm just not a fan of all that manual work of figuring out what the key sent to the server is, and then write a mapping to the active_record-style/undescored keys :-\

**Comment by beauby** [comment]
That's a legit point of view, although I believe users of AMS *should* know what kind of stuff is hitting their apps ;) Moreover, would one do the filtering with strong params, they would also have to deal with the client-issued data, and not some internal representation. Also, I tend to believe more and more deserialization does not really belong inside AMS in its current form (it is currently just a class unrelated to any other piece of code inside AMS).
At the end of the day, this is a performance vs convenience issue, and the most convenient way to handle it would probably be to have the client issue payloads such that no key has to be transformed.

**Comment by NullVoxPopuli** [comment]
I've actually thought about tweaking my ember's serializer again to go back to underscoring, rather than trying to achieve 100% recommended JSON API compliance. Clients have all this processing power that my poor lil server might not be utilizing to the fullest :-)

**Comment by beauby** [comment]
@NullVoxPopuli: Note that I just merged deserialization into `jsonapi_parser#master`, with a feature that may be of interest to you:
```ruby
options = {
  attributes: { only: [:blah1, :blah2] },
  relationships: { except: [:blah3, :blah4], polymorphic: [:blah5] },
  key_formatter: ->(x) { x.underscore }
}
JSON::API.parse(json_document).data.to_activerecord_hash(options)
```

**Comment by NullVoxPopuli** [comment]
that is of interest to me. :-)

that would solve my problem. I'll adopt your gem maybe in the next month or two. I'm slow on my project using AMS right now (lots of grunt work I'm avoiding)

But yeah, since the direction of AMS is to use your jsonapi_parser (name change pending), I'll go ahead and close this PR, in favor of using `key_formatter` in the future.

**Comment by NullVoxPopuli** [comment]
For those that have access to my repo, please don't delete the branch, I'm using it on my app. :-)
